src/snowflake/cli/api/metrics.py

```diff
@@ -213,15 +248,57 @@ def counters(self) -> Dict[str, int]:
         # return a copy of the original dict to avoid mutating the original
         return self._counters.copy()
 
+    @property
+    def num_spans_past_depth_limit(self) -> int:
+        return len(
+            [
+                span
+                for span in self._completed_spans
+                if span.span_depth > self.SPAN_DEPTH_LIMIT
+            ]
+        )
+
+    @property
+    def num_spans_past_total_limit(self) -> int:
+        return max(
+            0,
+            # we don't count the spans past the depth limit since they would not be reported
+            len(self._completed_spans)
+            - self.num_spans_past_depth_limit
+            - self.SPAN_TOTAL_LIMIT,
+        )
+
     @property
     def completed_spans(self) -> List[Dict]:
         """
-        returns the completed spans tracked throughout a command, sorted by start time, for reporting telemetry
+        Returns the completed spans tracked throughout a command, sorted by start time, for reporting telemetry
+
+        Ensures that the spans we send are within the configured limits and marks
+        certain spans as trimmed if their children would bypass the limits we set
         """
-        return [
-            step.to_dict()
-            for step in sorted(
-                self._completed_spans,
-                key=lambda step: step.start_time,
+        # take spans breadth-first within the depth and total limits
+        # since we care more about the big picture than granularity
+        spans_to_report = set(
+            nsmallest(
+                n=self.SPAN_TOTAL_LIMIT,
+                iterable=(
+                    span
+                    for span in self._completed_spans
+                    if span.span_depth <= self.SPAN_DEPTH_LIMIT
+                ),
+                key=lambda span: (span.span_depth, span.start_time),
             )
+        )
+
+        # sort by start time to make reading the payload easier
+        sorted_spans_to_report = sorted(
+            spans_to_report, key=lambda span: span.start_time
+        )
+
+        return [
+            {
+                **span.to_dict(),
+                CLIMetricsSpan.TRIMMED_KEY: not span.children <= spans_to_report,
```

Comment:
we can use the inverse operator to avoid the `not`. (there's `.issuperset()` but unfortunately it would return `Ture` if they're equal, and there's no named method for strict superset checking)
```suggestion
                CLIMetricsSpan.TRIMMED_KEY: span.children > spans_to_report,
```